doc/decisions/DR016-Related-entities-result-paging.md

```diff
@@ -0,0 +1,333 @@
+  i.e. when there are still pages remaining for some inputs but not
+  others.
+- Additional metadata in the object is limited to whole-batch data.
+
+
+## Options considered
+
+### Option 1 - Whole-batch page token
+
+Host supplies a batch of entity references and a callback. The callback
+is called for each input entity, with one of its arguments being a
+page of related entities as a simple list.
+
+Once callbacks have been called, the function returns a page token. This
+can be supplied to a subsequent call to get the next page of results for
+all entities in the input batch simultaneously.
+
+Note that this could also be designed to use a cursor object with
+similar pros and cons, but that option was discarded (see Background
+section).
+
+```python
+class AbstractPageToken:
+    ...
+
+SuccessCallback = Callable[[int, List[EntityReferences]], None]
+
+def relatedEntities(
+        batch: List[EntityReference],
+        relationship: TraitsData,
+        successCb: SuccessCallback,
+        errorCb,
+        limit: int = -1,
+        pageToken: AbstractPageToken = None) -> AbstractPageToken: ...
+```
+
+#### Pros
+
+- Conceptually simple.
+- Whole batch is paged at once, retaining batch-awareness in the
+  manager for subsequent pages.
+- Per-element callback means it's trivial to skip processing elements of
+  the input batch whose pages are exhausted.
+- Matches the most common pattern of pageable HTTP APIs allowing e.g.
+  the page token to be a thin wrapper around a "real" page token(s).
+- Retains existing design in line with OpenAssetIO design principles,
+  i.e. non-paged signature is simply the same function but ignoring
+  optional arguments.
+- Manager plugin can remain conceptually stateless, since required state
+  is bundled up in the page token.
+- Page token is (or can be) immutable, aiding thread-safety.
+
+#### Cons
+
+- Adding more functionality (e.g. previous page) requires additional
+  sibling functions or arguments.
+- Adding metadata (e.g. total results) requires additional sibling
+  functions or arguments.
+
+
+### Option 2 - Cursor objects / iterators
+
+Firstly let's consider the general pros and cons of object-oriented
+cursor objects.
+
+```python
+class AbstractPage:
+
+    def get(self) -> List[EntityReferences]:
+        # Current page of refs.
+        ...
+
+    def next(self) -> AbstractPage:
+        # Get next page.
+        ...
+
+    def remaining(self) -> Union[int, None]:
+        # None means "unsupported".
+        ...
+```
+
+#### Pros
+
+- Retrieving the data can be decoupled from traversal of pages, allowing
+  for lazy network retrieval of desired pages.
+- Additional metadata can be included in the cursor (e.g. remaining
+  pages) in an extensible way.
+
+#### Cons
+
+- Manager plugin must implement yet another interface.
+- Cursor objects are stateful by design, breaking the stateless
+  design principle of interactions with manager plugins. This causes
```

Comment:
Ah good point, when I wrote this I assumed a mutable cursor object, which does seem to mirror DB cursor objects - probably need to do some more research there.  

A mutable cursor object vs. an immutable page token is what I had in mind here, where sharing a mutable cursor across threads introduces thread-safety issues - the very issues `ManagerInterface` claims to avoid by being stateless.

But an immutable cursor object does seem to solve this problem :thinking:

Reply:
I've split the cons to hopefully cover both immutable and mutable cursor types: 73795d74

Reply:
Further updated the relevant pros and cons in cf837603